README.rst
============================
django-elasticsearch-dsl-drf
============================
Integrate `Elasticsearch DSL
<https://pypi.python.org/pypi/elasticsearch-dsl>`_ with
`Django REST framework <https://pypi.python.org/pypi/djangorestframework>`_ in
the shortest way possible, with least efforts possible.

Package provides views, serializers, filter backends, pagination and other
handy add-ons.

You are expected to use `django-elasticsearch-dsl
<https://pypi.python.org/pypi/django-elasticsearch-dsl>`_ for defining your
Elasticsearch documents.

.. image:: https://img.shields.io/pypi/v/django-elasticsearch-dsl-drf.svg
   :target: https://pypi.python.org/pypi/django-elasticsearch-dsl-drf
   :alt: PyPI Version

.. image:: https://img.shields.io/pypi/pyversions/django-elasticsearch-dsl-drf.svg
    :target: https://pypi.python.org/pypi/django-elasticsearch-dsl-drf/
    :alt: Supported Python versions

.. image:: https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/workflows/test/badge.svg
   :target: https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/actions
   :alt: Build Status

.. image:: https://readthedocs.org/projects/django-elasticsearch-dsl-drf/badge/?version=latest
    :target: http://django-elasticsearch-dsl-drf.readthedocs.io/en/latest/?badge=latest
    :alt: Documentation Status

.. image:: https://img.shields.io/badge/license-GPL--2.0--only%20OR%20LGPL--2.1--or--later-blue.svg
   :target: https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/#License
   :alt: GPL-2.0-only OR LGPL-2.1-or-later

.. image:: https://coveralls.io/repos/github/barseghyanartur/django-elasticsearch-dsl-drf/badge.svg?branch=master
    :target: https://coveralls.io/github/barseghyanartur/django-elasticsearch-dsl-drf?branch=master
    :alt: Coverage

Documentation
=============
Documentation is available on `Read the Docs
<http://django-elasticsearch-dsl-drf.readthedocs.io/>`_.

Make sure to read `FAQ <https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/blob/master/docs/faq.rst>`_.

Prerequisites
=============
- Django 3.2, 4.1 and 4.2.
- Python 3.8, 3.9, 3.10, 3.11 and 3.12.
- Elasticsearch 6.x, 7.x. For older versions use
  ``django-elasticsearch-dsl-drf`` version 0.18.

Main features and highlights
============================

- :doc:`Dynamic serializer for Documents <basic_usage_examples>`.
- :doc:`Search filter backend <advanced_usage_examples>`.
- :doc:`Ordering filter backend <advanced_usage_examples>`.
- :doc:`Filtering filter backend <advanced_usage_examples>` (big variety of
  native- and functional- query lookups, such as ``gt``, ``gte``, ``lt``,
  ``lte``, ``endswith``, ``contains``, ``wildcard``, ``exists``, ``exclude``,
  ``isnull``, ``range``, ``in``, ``prefix`` (same as ``startswith``), ``term``
  and ``terms`` is implemented.
- :doc:`Geo-spatial filtering filter backend <advanced_usage_examples>` (the
  following filters implemented: ``geo_distance``, ``geo_polygon`` and
  ``geo_bounding_box``).
- :doc:`Geo-spatial ordering filter backend <advanced_usage_examples>` (the
  following filters implemented: ``geo_distance``).
- :doc:`Faceted search filter backend <advanced_usage_examples>`.
- :doc:`Post-filter filter backend <advanced_usage_examples>`.
- :doc:`Nested filtering filter backend <nested_fields_usage_examples>`.
- :doc:`Highlight backend <advanced_usage_examples>`.
- :doc:`Suggester filter backend <advanced_usage_examples>`.
- :doc:`Functional suggester filter backend <advanced_usage_examples>`.
- :doc:`Pagination (Page number and limit/offset pagination) <advanced_usage_examples>`.
- :doc:`Ids filter backend <advanced_usage_examples>`.
- :doc:`Multi match search filter backend <search_backends>`.
- :doc:`Simple search query search filter backend <search_backends>`.
- :doc:`More-like-this support (detail action) <more_like_this>`.
- :doc:`Global aggregations support <global_aggregations>`.
- :doc:`Source filter backend <source_backend>`.

Do you need a similar tool for GraphQL? Check `graphene-elastic
<https://github.com/barseghyanartur/graphene-elastic>`__.

Demos
=====
Online demo
-----------
Check the `live demo app <https://django-elasticsearch-dsl-drf.herokuapp.com>`__
(Django 3.1 + Django REST Framework 3.12 + Elasticsearch 7.5) hosted on Heroku
and bonsai.io.

Local demo
----------
A frontend demo (React based) is available. See the `dedicated docs
<https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/blob/master/examples/frontend/README.rst>`_
for more information.

To bootstrap evaluation, clone the repository locally and run `docker-compose`.

.. code-block:: sh

    docker-compose up

It will set up:

- Elasticsearch on `http://localhost:9200 <http://localhost:9200>`_
- Django REST framework on `http://localhost:8000 <http://localhost:8000>`_
- React on `http://localhost:3000 <http://localhost:3000>`_

Installation
============
(1) Install latest stable version from PyPI:

    .. code-block:: sh

        pip install django-elasticsearch-dsl-drf

    or latest stable version from GitHub:

    .. code-block:: sh

        pip install https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/archive/stable.tar.gz

(2) Add ``rest_framework``, ``django_elasticsearch_dsl`` and
    ``django_elasticsearch_dsl_drf`` to ``INSTALLED_APPS``:

    .. code-block:: python

        INSTALLED_APPS = (
            # ...
            # REST framework
            'rest_framework',

            # Django Elasticsearch integration
            'django_elasticsearch_dsl',

            # Django REST framework Elasticsearch integration (this package)
            'django_elasticsearch_dsl_drf',
            # ...
        )

Quick start
===========
Perhaps the easiest way to get acquainted with ``django-elasticsearch-dsl-drf``
is to read the :doc:`quick start tutorial <quick_start>`.

See it as a guide of diving into integration of Elasticsearch with Django
with very low knowledge entry level.

Testing
=======
Project is covered with tests.

To test with all supported Python/Django versions type:

.. code-block:: sh

    tox

To test against specific environment, type:

.. code-block:: sh

    tox -e py38-django30

To test just your working environment type:

.. code-block:: sh

    ./runtests.py

To run a single test in your working environment type:

.. code-block:: sh

    ./runtests.py src/django_elasticsearch_dsl_drf/tests/test_filtering.py

Or:

.. code-block:: sh

    ./manage.py test django_elasticsearch_dsl_drf.tests.test_ordering

To run a single test class in a given test module in your working environment
type:

.. code-block:: sh

    ./runtests.py src/django_elasticsearch_dsl_drf/tests/test_suggesters.py::TestSuggesters

It's assumed that you have all the requirements installed. If not, first
install the test requirements:

.. code-block:: sh

    pip install -r examples/requirements/test.txt

Writing documentation
=====================
Keep the following hierarchy.

.. code-block:: text

    =====
    title
    =====

    header
    ======

    sub-header
    ----------

    sub-sub-header
    ~~~~~~~~~~~~~~

    sub-sub-sub-header
    ^^^^^^^^^^^^^^^^^^

    sub-sub-sub-sub-header
    ++++++++++++++++++++++

    sub-sub-sub-sub-sub-header
    **************************

License
=======
GPL-2.0-only OR LGPL-2.1-or-later

Support
=======
For any issues contact me at the e-mail given in the `Author`_ section.

Author
======
Artur Barseghyan <artur.barseghyan@gmail.com>


docs/changelog.rst
Release history and notes
=========================
`Sequence based identifiers
<http://en.wikipedia.org/wiki/Software_versioning#Sequence-based_identifiers>`_
are used for versioning (schema follows below):

.. code-block:: text

    major.minor[.revision]

- It's always safe to upgrade within the same minor version (for example, from
  0.3 to 0.3.4).
- Minor version changes might be backwards incompatible. Read the
  release notes carefully before upgrading (for example, when upgrading from
  0.3.4 to 0.4).
- All backwards incompatible changes are mentioned in this document.

unreleased
----------

- Adds support for Django 4.1 & 4.2
- Drops support for Django 2.2 & 3.1
- Adds support for Python 3.10, 3.11 & 3.12
- Drops support for Python 3.7

0.22.5
------
2022-07-04

- Return 0 for count when using EmptySearch.

0.22.4
------
2022-01-24

- Skip some parts of the ViewSet class definition on empty `document`.

0.22.3
------
2022-01-24

- Make running of ViewSet checks flexible (to allow dynamic document definition).
- Added the `FacetedFilterSearchFilterBackend`.

0.22.2
------
2021-08-29

- Tested against Django 3.2.

0.22.1
------
2021-05-01

- `DictionaryProxy` gets an optional `meta` argument, which will hold meta
  information of the hit.

0.22
----
2021-03-26

.. note::

    Release dedicated to my dear son, Raffi, who turns 16 at the end of the
    month. Happy birthday, dear Raffi.

- Make it easier to override the `DictionaryProxy` by moving it to the
  scope of the `BaseDocumentViewSet`.
- Add `tzinfo` check for better date vs datetime determination.
- Broader `EmptySearch` compatibility with `elasticsearch_dsl.Search`.
- Remove old Django code from docs.
- Allow more specific targeting for nested sort fields.
- Add more tests.

0.21
----
2021-02-04

.. note::

    Release dedicated to defenders of Armenia and Artsakh (Nagorno Karabakh)
    and all the victims of Turkish and Azerbaijani aggression.

- Drop support for Python 2.7 and 3.5.
- Drop support for Django 1.11, 2.0 and 2.1.
- Tested against Django 3.1 and Python 3.9.
- Take options (ex: boost, fuzziness) into consideration
  in `NestedQueryBackend`.
- Added an example of a proper date formatting.
- Added an example of alternative document ID.
- Added experimental `QueryFriendlyPageNumberPagination` which fires just a single
  query instead of used two used by `PageNumberPagination`.

0.20.9
------
2020-10-16

.. note::

    Help to rebuild the Armenian homeland. Please, `make a donation
    <https://www.himnadram.org/en>`__. All donations received are solely for
    humanitarian purposes.

- Implemented ``geo_shape`` filter.

0.20.8
------
2020-04-10

- Fixes in ``skip_duplicates`` option support for native suggester.

0.20.7
------
2020-04-10

- Make ``skip_duplicates`` available for native suggester.

0.20.6
------
2020-04-04

- Basic implementation of RegExp field lookup.

0.20.5
------
2019-12-30

- Minor fixes.

0.20.4
------
2019-12-25

- Tested against Django 3.0.
- Tested against Python 3.8.
- Tested against Django REST Framework 3.11.
- Minor fixes.
- Test optimisations.

0.20.3
------
2019-09-20

- Testing the auxiliary versions module.

0.20.2
------
2019-08-30

- Minor improvements in test coverage.

0.20.1
------
2019-08-18

- Minor Elasticsearch 7.x compatibility fixes.

0.20
----
2019-08-17

- Adding Elasticsearch 7.x support.

0.19
----
2019-08-06

.. note::

    Dropping support for Elasticsearch versions prior 6.x. This is unfortunate,
    but this project depends on the upstream ``django-elasticsearch-dsl`` where
    as of version 6.4.x the support for older Elasticsearch versions was
    dropped. Use ``django-elasticsearch-dsl-drf`` version 0.18 if you need
    to work with 5.x or 2.x.

- Dropping support for Elasticsearch versions prior to 6.x.

0.18
----
2019-06-26

.. note::

    Support for Django versions prior 1.11 has been dropped.
    Support for Django REST Framework prior 3.9 has been dropped.

- Dropping support for Django versions prior 1.11.
- Dropping support for Django REST Framework versions prior 3.9.
- Fix Django REST Framework deprecations.

0.17.7
------
2019-05-30

.. note::

    Support for Django 1.8, 1.9 and 1.10 will be dropped in the next release.
    As usual, compatibility shims won't be removed directly. The change
    will affect the test matrix only first.

- Prevent unicode errors in tests on Python 2.7.
- Fixes in occasionally failing search test (``test_search`` and
  ``test_filtering_geo_spatial``).
- Working travis.
- Fixed issue with errors on empty ``ids`` filter.

0.17.6
------
2019-04-08

- Minor fixes.
- Additions to the docs.

0.17.5
------
2019-04-03

.. note::

    Dropping support for Python 3.4. As of this version everything works, but
    no longer tested.

- Minor fixes.
- Dropping Python 3.4 support.
- Django 2.2 support.

0.17.4
------
2019-03-13

- Source backend.

0.17.3
------
2019-02-08

- Obey object permissions.

0.17.2
------
2019-01-07

- Add nested ordering.

0.17.1
------
2018-12-12

- Skipping the new context suggester tests for Elasticsearch 2.x and a number
  of other 2.x related fixes in tests.
- A number of 5.x fixes in tests.

0.17
----
2018-12-12

.. note::

    Release supported by `whythawk <https://github.com/whythawk>`_.

- Added support for context suggesters (`category` and `geo`). Note, that
  this functionality is available for Elasticsearch 5.x and 6.x (thus, not
  for Elasticsearch 2.x).
- Added support for `size` attribute on suggesters.

0.16.3
------
2018-10-31

.. note::

    Release dedicated to Charles Aznavour.

- Make it possible to ignore certain Elastic exceptions by providing the
  appropriate ``ignore`` argument (on the view level). Default behaviour is
  intact. Set it to a list of integers (error codes) if you need it so.

0.16.2
------
2018-09-21

- Tested yet untested ``pip_helpers`` module.
- More tests.

0.16.1
------
2018-09-18


examples/simple/settings/base.py
# Django settings for example project.
import os
import sys

from .core import PROJECT_DIR, gettext


# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
BASE_DIR = os.path.dirname(os.path.dirname(__file__))

DEBUG = False
DEBUG_TOOLBAR = False
DEV = False

ADMINS = (
    # ('Your Name', 'your_email@example.com'),
)

MANAGERS = ADMINS

DATABASES = {
    'default': {
        # Add 'postgresql_psycopg2', 'mysql', 'sqlite3' or 'oracle'.
        'ENGINE': 'django.db.backends.sqlite3',
        # Or path to database file if using sqlite3.
        'NAME': PROJECT_DIR(os.path.join('..', '..', 'db', 'example.db')),
        # The following settings are not used with sqlite3:
        'USER': '',
        'PASSWORD': '',
        # Empty for localhost through domain sockets or '127.0.0.1' for
        # localhost through TCP.
        'HOST': '',
        # Set to empty string for default.
        'PORT': '',
    }
}

# Hosts/domain names that are valid for this site; required if DEBUG is False
# See https://docs.djangoproject.com/en/1.5/ref/settings/#allowed-hosts
ALLOWED_HOSTS = ['*']

# Local time zone for this installation. Choices can be found here:
# http://en.wikipedia.org/wiki/List_of_tz_zones_by_name
# although not all choices may be available on all operating systems.
# In a Windows environment this must be set to your system time zone.
TIME_ZONE = 'America/Chicago'

# Language code for this installation. All choices can be found here:
# http://www.i18nguy.com/unicode/language-identifiers.html
LANGUAGE_CODE = 'en'

LANGUAGES = (
    ('en', gettext("English")),  # Main language!
    ('hy', gettext("Armenian")),
    ('nl', gettext("Dutch")),
    ('ru', gettext("Russian")),
    ('de', gettext("German")),
)

SITE_ID = 1

# If you set this to False, Django will make some optimizations so as not
# to load the internationalization machinery.
USE_I18N = True

# If you set this to False, Django will not use timezone-aware datetimes.
USE_TZ = True

# Absolute filesystem path to the directory that will hold user-uploaded files.
# Example: "/var/www/example.com/media/"
MEDIA_ROOT = PROJECT_DIR(os.path.join('..', '..', 'media'))

# URL that handles the media served from MEDIA_ROOT. Make sure to use a
# trailing slash.
# Examples: "http://example.com/media/", "http://media.example.com/"
MEDIA_URL = '/media/'

# Absolute path to the directory static files should be collected to.
# Don't put anything in this directory yourself; store your static files
# in apps' "static/" subdirectories and in STATICFILES_DIRS.
# Example: "/var/www/example.com/static/"
STATIC_ROOT = PROJECT_DIR(os.path.join('..', '..', 'static'))

# URL prefix for static files.
# Example: "http://example.com/static/", "http://static.example.com/"
STATIC_URL = '/static/'

# Additional locations of static files
STATICFILES_DIRS = (
    # Put strings here, like "/home/html/static" or "C:/www/django/static".
    # Always use forward slashes, even on Windows.
    # Don't forget to use absolute paths, not relative paths.
    # PROJECT_DIR(os.path.join('..', '..', 'media', 'static')),
)

# List of finder classes that know how to find static files in
# various locations.
STATICFILES_FINDERS = (
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
    # 'django.contrib.staticfiles.finders.DefaultStorageFinder',
)

# Make this unique, and don't share it with anybody.
SECRET_KEY = '97818c*w97Zi8a-m^1coRRrmurMI6+q5_kyn*)s@(*_Pk6q423'

try:
    from .local_settings import DEBUG_TEMPLATE
except Exception as err:
    DEBUG_TEMPLATE = False

try:
    from .local_settings import USE_CACHED_TEMPLATE_LOADERS
except Exception as err:
    USE_CACHED_TEMPLATE_LOADERS = False

if USE_CACHED_TEMPLATE_LOADERS:

    _TEMPLATE_LOADERS = [
        ('django.template.loaders.cached.Loader', (
            'django.template.loaders.filesystem.Loader',
            'django.template.loaders.app_directories.Loader',
            # 'django.template.loaders.eggs.Loader',
        )),
    ]
else:

    _TEMPLATE_LOADERS = [
        'django.template.loaders.filesystem.Loader',
        'django.template.loaders.app_directories.Loader',
        # 'django.template.loaders.eggs.Loader',
    ]

TEMPLATE_DEBUG = DEBUG_TEMPLATE

# List of callables that know how to import templates from various
# sources.
TEMPLATE_LOADERS = _TEMPLATE_LOADERS

TEMPLATE_CONTEXT_PROCESSORS = (
    "django.contrib.auth.context_processors.auth",
    "django.core.context_processors.debug",
    "django.core.context_processors.i18n",
    "django.core.context_processors.media",
    "django.core.context_processors.static",
    "django.core.context_processors.tz",
    "django.contrib.messages.context_processors.messages",
    "django.core.context_processors.request",
)

TEMPLATE_DIRS = (
    # Put strings here, like "/home/html/django_templates" or
    # "C:/www/django/templates".
    # Always use forward slashes, even on Windows.
    # Don't forget to use absolute paths, not relative paths.
    PROJECT_DIR(os.path.join('..', 'templates')),
)

_MIDDLEWARE = (
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.locale.LocaleMiddleware',
    'corsheaders.middleware.CorsMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    # Uncomment the next line for simple clickjacking protection:
    # 'django.middleware.clickjacking.XFrameOptionsMiddleware',
)

ROOT_URLCONF = 'urls'

# Python dotted path to the WSGI application used by Django's runserver.
WSGI_APPLICATION = 'wsgi.application'

# FIXTURE_DIRS = (
#   PROJECT_DIR(os.path.join('..', 'fixtures'))
# )

INSTALLED_APPS = [
    # Django core and contrib apps
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.sites',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'django.contrib.admin',
    'django.contrib.sitemaps',
    'django.contrib.humanize',

    # Third party apps
    'rest_framework',  # REST framework
    'django_elasticsearch_dsl',  # Elasticsearch integration
    'corsheaders',  # For React demo

    # This app
    'django_elasticsearch_dsl_drf',

    # Other project specific apps
    'assets',  # Static files
    'books',  # Test app
    'search_indexes',  # Search app
]

REST_FRAMEWORK = {
    'DEFAULT_AUTHENTICATION_CLASSES': (
        'rest_framework.authentication.BasicAuthentication',
        'rest_framework.authentication.SessionAuthentication',
    ),
    'DEFAULT_PAGINATION_CLASS':
        'rest_framework.pagination.PageNumberPagination',
    'PAGE_SIZE': 100,
    'ORDERING_PARAM': 'ordering',
}

# Elasticsearch configuration
ELASTICSEARCH_DSL = {
    'default': {
        'hosts': 'localhost:9200',
        'timeout': 30,
    },
}

# CORS headers config
CORS_ORIGIN_ALLOW_ALL = True

# Name of the Elasticsearch index
ELASTICSEARCH_INDEX_NAMES = {
    'search_indexes.documents.address': 'address',
    'search_indexes.documents.author': 'author',
    'search_indexes.documents.book': 'book',
    'search_indexes.documents.tag': 'tag',
    'search_indexes.documents.city': 'city',
    'search_indexes.documents.journal': 'journal',
    'search_indexes.documents.publisher': 'publisher',
    'search_indexes.documents.location': 'location',
}

ELASTICSEARCH_DATETIME_FORMAT = '%Y-%m-%dT%H:%M:%SZ'

LOGIN_URL = '/en/accounts/login/'
LOGIN_ERROR_URL = '/en/accounts/login/'
LOGOUT_URL = '/en/accounts/logout/'
LOGIN_REDIRECT_URL = '/en/'

# A sample logging configuration. The only tangible logging
# performed by this configuration is to send an email to
# the site admins on every HTTP 500 error when DEBUG=False.
# See http://docs.djangoproject.com/en/dev/topics/logging for
# more details on how to customize your logging configuration.
LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'filters': {
        'require_debug_false': {
            '()': 'django.utils.log.RequireDebugFalse'
        }
    },
    'root': {
        'level': 'INFO',
        'handlers': ['all_log'],
    },
    'formatters': {
        'verbose': {
            'format': '\n%(levelname)s %(asctime)s [%(pathname)s:%(lineno)s] '
                      '%(message)s'
        },
        'simple': {
            'format': '\n%(levelname)s %(message)s'
        },
    },
    'handlers': {
        'mail_admins': {
            'level': 'ERROR',
            'filters': ['require_debug_false'],
            'class': 'django.utils.log.AdminEmailHandler'
        },
        'console': {
            'level': 'DEBUG',
            'class': 'logging.StreamHandler',
            'formatter': 'verbose'
        },
        'all_log': {
            'level': 'DEBUG',
            'class': 'logging.handlers.RotatingFileHandler',
            'filename': PROJECT_DIR("../../logs/all.log"),
            'maxBytes': 1048576,
            'backupCount': 99,
            'formatter': 'verbose',
        },
        'django_log': {
            'level': 'DEBUG',
            'class': 'logging.handlers.RotatingFileHandler',
            'filename': PROJECT_DIR("../../logs/django.log"),
            'maxBytes': 1048576,
            'backupCount': 99,
            'formatter': 'verbose',
        },
        'django_request_log': {
            'level': 'DEBUG',
            'class': 'logging.handlers.RotatingFileHandler',
            'filename': PROJECT_DIR("../../logs/django_request.log"),
            'maxBytes': 1048576,
            'backupCount': 99,
            'formatter': 'verbose',
        },
        'django_elasticsearch_dsl_drf_log': {
            'level': 'DEBUG',
            'class': 'logging.handlers.RotatingFileHandler',
            'filename': PROJECT_DIR("../../logs/debug_toolbar_force.log"),
            'maxBytes': 1048576,
            'backupCount': 99,
            'formatter': 'verbose',
        },
    },
    'loggers': {
        'django.request': {
            'handlers': ['django_request_log'],
            'level': 'INFO',
            'propagate': True,
        },
        'django': {
            'handlers': ['django_log'],
            'level': 'ERROR',
            'propagate': False,
        },
        'books': {
            'handlers': ['console', 'django_elasticsearch_dsl_drf_log'],
            'level': 'DEBUG',
            'propagate': True,
        },
        'django_elasticsearch_dsl_drf': {
            'handlers': ['console', 'django_elasticsearch_dsl_drf_log'],
            'level': 'DEBUG',
            'propagate': True,
        }
    },
}

# Do not put any settings below this line
try:
    from .local_settings import *
except ImportError:
    pass

if DEBUG and DEBUG_TOOLBAR:
    try:
        # Make sure the django-debug-toolbar is installed
        import debug_toolbar

        # debug_toolbar
        _MIDDLEWARE += (
            'debug_toolbar.middleware.DebugToolbarMiddleware',
            'debug_toolbar_force.middleware.ForceDebugToolbarMiddleware',
        )

        INSTALLED_APPS += (
            'debug_toolbar',
            'elastic_panel',
        )

        DEBUG_TOOLBAR_CONFIG = {
            'INTERCEPT_REDIRECTS': False,
            'JQUERY_URL': STATIC_URL + 'js/vendor/jquery.js',
        }

        DEBUG_TOOLBAR_PANELS = (
            # Defaults
            # 'debug_toolbar.panels.versions.VersionsPanel',
            'debug_toolbar.panels.timer.TimerPanel',
            'debug_toolbar.panels.settings.SettingsPanel',
            'debug_toolbar.panels.headers.HeadersPanel',
            'debug_toolbar.panels.request.RequestPanel',
            'debug_toolbar.panels.sql.SQLPanel',
            # 'debug_toolbar.panels.staticfiles.StaticFilesPanel',
            # 'debug_toolbar.panels.templates.TemplatesPanel',
            'debug_toolbar.panels.cache.CachePanel',
            'debug_toolbar.panels.signals.SignalsPanel',
            'debug_toolbar.panels.logging.LoggingPanel',
            'debug_toolbar.panels.redirects.RedirectsPanel',
            # Additional
            'elastic_panel.panel.ElasticDebugPanel',
        )

    except ImportError:
        pass

# Make the `django-elasticsearch-dsl-drf` package available without
# installation or development.
if DEV:
    app_source_path = os.environ.get(
        'DJANGO_ELASTICSEARCH_DSL_DRF_SOURCE_PATH',
        'src'
    )
    # sys.path.insert(0, os.path.abspath('src'))
    sys.path.insert(0, os.path.abspath(app_source_path))

MIDDLEWARE_CLASSES = _MIDDLEWARE
